crates/paimon/src/spec/tokens.rs

```diff
@@ -0,0 +1,1246 @@
+        DataType::LocalZonedTimestamp(LocalZonedTimestampType::new(precision).unwrap())
+    }
+
+    pub fn parse_optional_precision(&mut self, default_precision: u32) -> u32 {
+        let mut precision = default_precision;
+        if self.has_next_token_type(vec![TokenType::BeginParameter]) {
+            self.next_token_type(TokenType::BeginParameter);
+            self.next_token_type(TokenType::LiteralInt);
+            precision = self.token_as_int() as u32;
+            self.next_token_type(TokenType::EndParameter);
+        }
+        precision
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct DataTypeJsonParser {}
+
+impl DataTypeJsonParser {
+    pub fn is_delimiter(&mut self, character: char) -> bool {
+        character.is_whitespace()
+            || character == CHAR_BEGIN_SUBTYPE
+            || character == CHAR_END_SUBTYPE
+            || character == CHAR_BEGIN_PARAMETER
+            || character == CHAR_END_PARAMETER
+            || character == CHAR_LIST_SEPARATOR
+            || character == CHAR_DOT
+    }
+
+    pub fn is_digit(&mut self, c: char) -> bool {
+        c.is_ascii_digit()
+    }
+
+    pub fn consume_escaped(
+        &mut self,
+        builder: &mut String,
+        chars: &str,
+        mut cursor: usize,
+        delimiter: char,
+    ) -> usize {
+        // skip delimiter
+        cursor += 1;
+
+        while cursor < chars.len() {
+            let cur_char = chars.chars().nth(cursor).unwrap();
+            if cur_char == delimiter
+                && cursor + 1 < chars.len()
+                && chars.chars().nth(cursor + 1).unwrap() == delimiter
+            {
+                // escaping of the escaping char e.g. "'Hello '' World'"
+                cursor += 1;
+                builder.push(cur_char);
+            } else if cur_char == delimiter {
+                break;
+            } else {
+                builder.push(cur_char);
+            }
+            cursor += 1;
+        }
+        cursor
+    }
+
+    pub fn consume_int(&mut self, builder: &mut String, chars: &str, mut cursor: usize) -> usize {
+        while cursor < chars.len() && self.is_digit(chars.chars().nth(cursor).unwrap()) {
+            builder.push(chars.chars().nth(cursor).unwrap());
+            cursor += 1;
+        }
+        cursor - 1
+    }
+
+    pub fn consume_identifier(
+        &mut self,
+        builder: &mut String,
+        chars: &str,
+        mut cursor: usize,
+    ) -> usize {
+        while cursor < chars.len() && !self.is_delimiter(chars.chars().nth(cursor).unwrap()) {
+            builder.push(chars.chars().nth(cursor).unwrap());
+            cursor += 1;
+        }
+        cursor - 1
+    }
+
+    pub fn tokenize(&mut self, chars: &str) -> Vec<Token> {
+        let mut tokens = Vec::new();
+        let mut builder = String::new();
+        let mut cursor = 0;
+        while cursor < chars.len() {
+            let cur_char = chars.chars().nth(cursor).unwrap();
```

Comment:
Repeatedly calling chars.chars().nth(cursor) to get the current character can lead to performance issues in Rust, especially when dealing with large strings. The reason is that each call to chars.chars().nth(cursor) iterates from the beginning of the string until it reaches the cursor position. This means that as the cursor increases, this operation becomes increasingly expensive.